Proposal to add `Hashable` conformance to `Async(Throwing)Stream.Continuation` (#2700)
swiftlang/swift-evolution: proposals/NNNN-async-stream-continuation-hashable-conformance.md

> # Feature name

Reviewer: ```suggestion
# `Hashable` conformance for `Async(Throwing)Stream.Continuation`
```

* Proposal: [SE-NNNN](NNNN-async-stream-continuation-hashable-conformance.md)
* Authors: [Mykola Pokhylets](https://github.com/nickolas-pohilets)
* Review Manager: TBD
* Status: **Awaiting review**
* Implementation: [swiftlang/swift#79457](https://github.com/swiftlang/swift/pull/79457)
* Review: ([pitch](https://forums.swift.org/t/pitch-add-hashable-conformance-to-asyncstream-continuation/77897))

## Introduction

> This proposal suggests to add `Hashable` conformance to `Async(Throwing)Stream.Continuation`

Reviewer: ```suggestion
This proposal adds a `Hashable` conformance to `Async(Throwing)Stream.Continuation`
```

to simplify working with multiple streams.

## Motivation

> Use cases operating with multiple `AsyncStream`'s may need to store multiple continuations.

Reviewer: ```suggestion
Use cases operating with multiple `AsyncStream`s may need to store multiple continuations.
```

> When handling `onTermination` callback, client code needs to remove relevant continuation.

Reviewer: ```suggestion
When handling `onTermination` callback, client code needs to remove the relevant continuation.
```

> To identify relevant continuation, client code needs to be able to compare continuations.

Reviewer: ```suggestion
To identify the relevant continuation, client code needs to be able to compare continuations.
```

> It is possible to associate lookup key with each continuation, but this is inefficient.

Reviewer: ```suggestion
It is possible to associate a lookup key with each continuation, but this is inefficient.
```

`AsyncStream.Continuation` already stores a reference to `AsyncStream._Storage`,
whose identity can be used to provide simple and efficient `Hashable` conformance.

> Consider this simple Observer pattern with `AsyncSequence`-based API.

Reviewer: ```suggestion
Consider this simple Observer pattern with an `AsyncSequence`-based API.
```

To avoid implementing `AsyncSequence` from scratch it uses `AsyncStream` as a building block.

> To support multiple subscribes new stream is returned every time.

Reviewer: ```suggestion
To support multiple subscribers, a new stream is returned every time.
```

```swift
@MainActor private class Sender {
    var value: Int = 0 {
        didSet {
            for c in continuations {
                c.yield(value)
            }
        }
    }

    var values: some AsyncSequence<Int, Never> {
        AsyncStream<Int>(bufferingPolicy: .bufferingNewest(1)) { continuation in
            continuation.yield(value)
            self.continuations.insert(continuation)
            continuation.onTermination = { _ in
                DispatchQueue.main.async {
                    self.continuations.remove(continuation)
                }
            }
        }
    }

    private var continuations: Set<AsyncStream<Int>.Continuation> = []
}
```

> Without `Hashable` conformance, continuation needs to be associated with an artificial identifier,

Reviewer: ```suggestion
Without a `Hashable` conformance, each continuation needs to be associated with an artificial identifier,
```

> e.g. object identity by wrapping each continuation in a class:

Reviewer: ```suggestion
(e.g., object identity) by wrapping each continuation in a class:
```

Author: "wrapping" is still part of the example. Wrapping in a class is not required, there are alternatives, e.g. one could store tuples of `(continuation, UUID)`.

```swift
@MainActor private class Sender {
    var value: Int = 0 {
        didSet {
            for c in continuations {
                c.value.yield(value)
            }
        }
    }

    var values: some AsyncSequence<Int, Never> {
        AsyncStream<Int> { (continuation: AsyncStream<Int>.Continuation) -> Void in
            continuation.yield(value)
            let box = ContinuationBox(value: continuation)
            self.continuations.insert(box)
            continuation.onTermination = { _ in
                DispatchQueue.main.async {
                    self.continuations.remove(box)
                }
            }
        }
    }

    private var continuations: Set<ContinuationBox> = []

    private final class ContinuationBox: Hashable, Sendable {
        let value: AsyncStream<Int>.Continuation

        init(value: AsyncStream<Int>.Continuation) {
            self.value = value
        }

        static func == (lhs: Sender.ContinuationBox, rhs: Sender.ContinuationBox) -> Bool {
            lhs === rhs
        }

        func hash(into hasher: inout Hasher) {
            hasher.combine(ObjectIdentifier(self))
        }
    }
}
```

> Note that capturing `continuation` or `box` in `onTermination` is safe, because `onTermination` is dropped after being called.

Reviewer: ```suggestion
Note that capturing `continuation` or `box` in `onTermination` is safe, because `onTermination` is dropped after being called
```

> And it is always called, even if `AsyncStream` is discarded without even being iterated.

Reviewer: ```suggestion
(and it is _always_ called, even if `AsyncStream` is discarded without being iterated).
```

## Proposed solution

> Add `Hashable` conformance to `Async(Throwing)Stream.Continuation`.

Reviewer: ```suggestion
Add a `Hashable` conformance to `Async(Throwing)Stream.Continuation`.
```

## Detailed design

> Every time when `build` closure of the `Async(Throwing)Stream.init()` is called,

Reviewer: ```suggestion
Every time when the `build` closure of the `Async(Throwing)Stream.init()` is called,
```

it receives a continuation distinct from all other continuations.
All copies of the same continuation should compare equal.

> Yielding values or errors, finishing the stream or cancelling iteration should not affect equality.

Reviewer: ```suggestion
Yielding values or errors, finishing the stream, or cancelling iteration should not affect equality.
```

Assigning `onTermination` closures should not affect equality.

## Source compatibility

This is an additive change.

Retroactive conformances are unlikely to exist, because current public API of the `Async(Throwing)Stream.Continuation` 
does not provide anything that could be reasonably used to implement `Hashable` or `Equatable` conformances.

## ABI compatibility

This is an additive change. 

## Implications on adoption

Adopters will need a new version of the standard library.

## Alternatives considered

> ### Backdeployment

Reviewer: Back-deployment is generally a platform-level concern anyway (i.e., at the platform owner's discretion) and so not typically considered as part of proposals.

> It is possible to use `@backDeployed` on functions implementing protocol requirements, but it is not possible to back-deploy conformance record.

Reviewer: ```suggestion
It is possible to use `@backDeployed` on functions implementing protocol requirements, but it is not possible to back-deploy the conformance record itself.
```
